src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -2362,148 +2392,115 @@ private void InferTypeFrom(VariableExpressionAst variableExpressionAst, List<PST
                 }
-            }
-            else
-            {
-                inferredTypes.Add(new PSTypeName(_context.CurrentTypeDefinitionAst));
-                return;
-            }
 
-            if (parent is null)
-            {
                 return;
             }
 
-            // Look for our variable as a parameter or on the lhs of an assignment - hopefully we'll find either
-            // a type constraint or at least we can use the rhs to infer the type.
-            while (parent.Parent != null)
+            for (int i = 0; i < SpecialVariables.AutomaticVariables.Length; i++)
             {
-                parent = parent.Parent;
-            }
-
-            int startOffset = variableExpressionAst.Extent.StartOffset;
-            var targetAsts = (List<Ast>)AstSearcher.FindAll(
-                parent,
-                ast =>
+                if (!astVariablePath.UnqualifiedPath.EqualsOrdinalIgnoreCase(SpecialVariables.AutomaticVariables[i]))
                 {
-                    if (ast is ParameterAst || ast is AssignmentStatementAst || ast is CommandAst)
-                    {
-                        return variableExpressionAst.AstAssignsToSameVariable(ast)
-                            && ast.Extent.EndOffset < startOffset;
-                    }
+                    continue;
+                }
 
-                    if (ast is ForEachStatementAst)
-                    {
-                        return variableExpressionAst.AstAssignsToSameVariable(ast)
-                            && ast.Extent.StartOffset < startOffset;
-                    }
+                Type type = SpecialVariables.AutomaticVariableTypes[i];
+                if (type != typeof(object))
+                {
+                    inferredTypes.Add(new PSTypeName(type));
+                }
 
-                    return false;
-                },
-                searchNestedScriptBlocks: true);
+                return;
+            }
 
-            foreach (var ast in targetAsts)
+            // This visitor + loop finds the start of the current scope and traverses top to bottom to find the nearest variable assignment.
+            // Then repeats the process for each parent scope.
+            var assignmentVisitor = new VariableAssignmentVisitor()
+            {
+                ScopeIsLocal = true,
+                LocalScopeOnly = variableExpressionAst.VariablePath.IsLocal || variableExpressionAst.VariablePath.IsPrivate,
+                StopSearchOffset = variableExpressionAst.Extent.StartOffset,
+                VariableTarget = variableExpressionAst
+            };
+            while (parent is not null)
             {
-                if (ast is ParameterAst parameterAst)
+                if (parent is IParameterMetadataProvider)
                 {
-                    var currentCount = inferredTypes.Count;
-                    inferredTypes.AddRange(InferTypes(parameterAst));
+                    parent.Visit(assignmentVisitor);
```

Comment:
Will this rescan already processed low level scopes? If yes, makes sense to complicate the code to stop this?

Reply:
The `StartOffset` of the last scanned scope is already used as the `StopSearchOffset` so we already have the logic to stop it.

Reply:
Isn't the parent extent wider than the child one?

Reply:
Your comment made me realize I had made a mistake in the scope logic that I've now fixed, so thank you for that.
As for the question, consider this scenario:
```
function Outer
{
    [string] $Test = "Hello"
    function Inner
    {
        $Test.<Tab>
    }
    $Test = Get-Acl
}
```
We are trying to infer the value of `$Test` inside `Inner`. We first search the inner scope and find no assignments so we move on to the `Outer` scope after setting the `StopSearchOffset` to the beginning of the previous scope (`Inner`). As a result, it will see the string assignment, then it will hit the function definition and stop the search because that's the stop offset. As a result it correctly infers that the type is a string, and not an ACL.